app/scripts/models/reliers/oauth.js

```diff
@@ -204,6 +207,15 @@ define(function (require, exports, module) {
     },
 
     /**
+     * Check if the relier wants access to the account encryption keys.
+     *
+     * @returns {Boolean}
+     */
+    wantsKeys () {
+      return !! (this.has('keysJwk') && this._config.scopedKeysEnabled);
```

Comment:
Some things that we could profitably check, either here or in a preceding method such as `fetch`:

* If the relier as provided `keys_jwk`, then:
  * are they one of the client-ids that's currently allowed to request keys?
  * are they requesting the scopes we expect them to request?
  * are they providing a redirect_uri at a known-good origin?

I know the oauth-server checks all those things already, but it is only able to check them pretty late in the game, after we've already authed the user and fetched their account keys.  If we're able to check some of these earlier without things getting too ugly, I think it would be a win, especially while it's early days for this code.  @shane-tomlinson @vladikoff thoughts?

Reply:
It seems like the sort of gating requested should be done on the oauth-server when requesting the key material, any client side gates can be worked around.

Reply:
@shane-tomlinson yea we already have oauth-server gates, but I think the security teams are pushing for extra front-end validation for v1.

Reply:
The plan is to have a 1:1 mapping of scopes and clients (for Lockbox, Notes)

Reply:
> @shane-tomlinson yea we already have oauth-server gates, but I think the security teams are pushing for extra front-end validation for v1.

I wasn't there in the meeting, so I have to ask extra questions. What does gating on the front end provide that gating on the back end doesn't?

Reply:
> It seems like the sort of gating requested should be done on the oauth-server
> when requesting the key material

By the time we ask the oauth-server for the extra key material to generate scoped keys, we've already fetched `kB` from the auth-server.  We could add extra gating in the auth-server I suppose, so that we can check it earlier.

> any client side gates can be worked around

Well, that depends on where the attacker is relative to the user...

Client-side gating would be defence-in-depth to protect against any issues in our server implementation, e.g. if there's a bug in oauth-server that let an attacker specify a different redirect_uri for the lockbox app, we have an additional opportunity to ensure the keys aren't being sent outside of Mozilla.  If someone's able to get a shell on one of the oauth-server webheads, a client-side check would prevent them from just silently redirecting all the keys to a URL they control.  Things like that.

That said, if you guys feel like it's not worth it at this stage, we can probably make the argument to defer that discussion and ship without it for lockbox.  It will become more of an issue if and when we add the ability to request sync encryption keys through this flow.

Reply:
from mtg: "check the redirect uris"++